### PR TITLE
incremental precompile correctness issues

### LINKED ISSUE #21141: World age failure when running GtkReactive demo
When running the `examples/drawing.jl` demo under
```julia
julia> versioninfo()
Julia Version 0.6.0-pre.alpha.229
Commit cc917b5 (2017-03-22 21:53 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-5500U CPU @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
```
and **if** GtkReactive needs to be rebuilt, I get the following error when I click on the canvas:
```julia
julia> include("drawing.jl")
INFO: Recompiling stale cache file /home/tim/.julia/lib/v0.6/GtkReactive.ji for module GtkReactive.
Gtk.GtkWindowLeaf(name="", parent, width-request=-1, height-request=-1, visible=TRUE, sensitive=TRUE, app-paintable=FALSE, can-focus=FALSE, has-focus=FALSE, is-focus=FALSE, can-default=FALSE, has-default=FALSE, receives-default=FALSE, composite-child=FALSE, style, events=0, no-show-all=FALSE, has-tooltip=FALSE, tooltip-markup=NULL, tooltip-text=NULL, window, opacity=1.000000, double-buffered, halign=GTK_ALIGN_FILL, valign=GTK_ALIGN_FILL, margin-left, margin-right, margin-start=0, margin-end=0, margin-top=0, margin-bottom=0, margin=0, hexpand=FALSE, vexpand=FALSE, hexpand-set=FALSE, vexpand-set=FALSE, expand=FALSE, scale-factor=1, border-width=0, resize-mode, child, type=GTK_WINDOW_TOPLEVEL, title="Drawing", role=NULL, resizable=TRUE, modal=FALSE, window-position=GTK_WIN_POS_NONE, default-width=-1, default-height=-1, destroy-with-parent=FALSE, hide-titlebar-when-maximized=FALSE, icon, icon-name=NULL, screen, type-hint=GDK_WINDOW_TYPE_HINT_NORMAL, skip-taskbar-hint=FALSE, skip-pager-hint=FALSE, urgency-hint=FALSE, accept-focus=TRUE, focus-on-map=TRUE, decorated=TRUE, deletable=TRUE, gravity=GDK_GRAVITY_NORTH_WEST, transient-for, attached-to, has-resize-grip, resize-grip-visible, application, is-active=TRUE, has-toplevel-focus=TRUE, startup-id, mnemonics-visible=FALSE, focus-visible=FALSE, is-maximized=FALSE)

julia> Failed to push!
    GtkReactive.MouseButton{GtkReactive.UserUnit}(GtkReactive.XY{GtkReactive.UserUnit}(GtkReactive.UserUnit(0.208377685546875), GtkReactive.UserUnit(0.3292300415039063)), 0x00000001, 4, 0)
to node
    Signal{GtkReactive.MouseButton{GtkReactive.UserUnit}}(GtkReactive.MouseButton{GtkReactive.UserUnit}(GtkReactive.XY{GtkReactive.UserUnit}(GtkReactive.UserUnit(0.208377685546875), GtkReactive.UserUnit(0.3292300415039063)), 0x00000001, 4, 0), nactions=1)
MethodError: no method matching (::##3#4)(::GtkReactive.MouseButton{GtkReactive.UserUnit})
The applicable method may be too new: running in world age 21622, while current world is 21627.
Closest candidates are:
  #3(::Any) at /home/tim/.julia/v0.6/GtkReactive/examples/drawing.jl:13 (method too new to be called from this world context.)
(::Reactive.##10#11{##3#4,Tuple{Reactive.Signal{GtkReactive.MouseButton{GtkReactive.UserUnit}}}})(::Reactive.Signal{Void}, ::Int64) at /home/tim/.julia/v0.6/Reactive/src/operators.jl:39
do_action(::Reactive.Action, ::Int64) at /home/tim/.julia/v0.6/Reactive/src/core.jl:141
run(::Int64) at /home/tim/.julia/v0.6/Reactive/src/core.jl:218
macro expansion at /home/tim/.julia/v0.6/Reactive/src/core.jl:258 [inlined]
(::Reactive.##7#8)() at ./task.jl:335
Failed to push!
    GtkReactive.MouseButton{GtkReactive.UserUnit}(GtkReactive.XY{GtkReactive.UserUnit}(GtkReactive.UserUnit(0.1935784912109375), GtkReactive.UserUnit(0.7561200714111328)), 0x00000001, 7, 256)
to node
    Signal{GtkReactive.MouseButton{GtkReactive.UserUnit}}(GtkReactive.MouseButton{GtkReactive.UserUnit}(GtkReactive.XY{GtkReactive.UserUnit}(GtkReactive.UserUnit(0.1935784912109375), GtkReactive.UserUnit(0.7561200714111328)), 0x00000001, 7, 256), nactions=1)
MethodError: no method matching (::##7#8)(::GtkReactive.MouseButton{GtkReactive.UserUnit})
The applicable method may be too new: running in world age 21622, while current world is 21627.
Closest candidates are:
  #7(::Any) at /home/tim/.julia/v0.6/GtkReactive/examples/drawing.jl:25 (method too new to be called from this world context.)
(::Reactive.##10#11{##7#8,Tuple{Reactive.Signal{GtkReactive.MouseButton{GtkReactive.UserUnit}}}})(::Reactive.Signal{Void}, ::Int64) at /home/tim/.julia/v0.6/Reactive/src/operators.jl:39
do_action(::Reactive.Action, ::Int64) at /home/tim/.julia/v0.6/Reactive/src/core.jl:141
run(::Int64) at /home/tim/.julia/v0.6/Reactive/src/core.jl:218
macro expansion at /home/tim/.julia/v0.6/Reactive/src/core.jl:258 [inlined]
(::Reactive.##7#8)() at ./task.jl:335
```

However, it seems to work fine if GtkReactive didn't need to be recompiled.

For you convenience I did a `egrep "pure|generated|eval" *` in every package that gets loaded for that demo, looking for language constructs that might be suspicious; results are in [this gist](https://gist.github.com/timholy/beaf63ec8f97dd30ba50f82ba644339a).

### PR DESCRIPTION
**Pull request by vtjnash**
This should fix #21141, and includes some other correctness improvements that I've known were missing, but require rather convoluted situations to manage to trigger them.

### CODE REVIEW

**Comment by tkelman** [comment]
can any of the issues being fixed here be tested?

**Comment by timholy** [comment]
I tested this branch; is it supposed to work with, or without, the `eval` change(s) to Reactive? With unmodified Reactive, I'm still getting world-age errors when I run the GtkReactive tests.

**Comment by vtjnash** [comment]
I'm only addressing the assertion failure. Reactive will need to be fixed separately.

**Comment by timholy** [comment]
Is it going to have to stay `eval`? That will presumably add quite a performance hit.

I can work on the changes to Reactive, but I don't want to apply a bandaid if a deeper fix is necessary/coming.

**Comment by vtjnash** [comment]
This is a bug in reactive. It needs to avoid starting its event loop before finishing loading code.

File: src/dump.c [hunk 1]
```diff
@@ -2123,33 +2164,70 @@ static void jl_insert_methods(linkedlist_t *list)
...
+                                    m = entry->func.method;
+                            }
+                            if (m != callee_mi->def) {
+                                // probably no good, just invalidate everything now
+                                jl_method_instance_delete(caller);
+                                break;
+                            }
+                        }
+                        if (callee_mi->max_world == ~(size_t)0) {
+                            // if this callee is still valid, just add a backedege
```
Comment (tkelman) [inline src/dump.c]: backedge misspelled

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: src/dump.c [hunk 2]
```diff
@@ -2999,26 +3087,72 @@ static void jl_update_backref_list(jl_value_t *old, jl_value_t *_new, size_t sta
...
+    size_t i, l = dependent_worlds->len;
+    for (i = 0; i < l; i++) {
+        size_t depworld = (size_t)dependent_worlds->items[i];
+        if (depworld <= world)
+            return depworld;
+    }
+    return jl_main_module->primary_world;
+}
+
+// repeated look up older methods until we come to one that existed
```
Comment (tkelman) [inline src/dump.c]: should this be "repeatedly look up" ?

File: test/compile.jl [hunk 3]
```diff
@@ -109,16 +122,39 @@ try
...
           end
           """)
     @test_throws ErrorException Core.kwfunc(Base.nothing) # make sure `nothing` didn't have a kwfunc (which would invalidate the attempted test)
 
     # Issue #12623
     @test __precompile__(true) === nothing
 
+    # Issue #21307
+    Base.require(Foo2_module)
+    @eval let Foo2_module = $(QuoteNode(Foo2_module)) # use @eval to see the results of loading the compile
```
Comment (tkelman) [inline test/compile.jl]: supposed to be a comma here?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21307#issuecomment-293070722) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/478ad12f8b6967b1deb6b1acc39e1b49854cbd57/5316064_vs_33aa319/report.md). cc @jrevels

**Comment by timholy** [comment]
> This is a bug in reactive. It needs to avoid starting its event loop before finishing loading code.

I'm still wrapping my head around this, but if I understand your suggestion you're not talking about Reactive's code (`__init__` runs at the end of module initialization, right?), but code that's defined after Reactive is loaded. If so, I suspect this suggestion would be limiting, more so than the `eval` trick. Let's say that Reactive didn't launch its own event queue, and left that up to user code. Once you started working with a Reactive GUI at the REPL (necessitating that you launch the event queue), that REPL becomes "frozen" and can't have new GUIs defined. I could easily be misunderstanding, however.

Perhaps this is an unworkable/stupid idea, but what about defining a compile hint `::LATEST`, aka
```julia
function connect_map(f::LATEST, output, inputs...)
    let prev_timestep = 0
        for inp in inputs
            add_action!(inp, output) do output, timestep
                if prev_timestep != timestep
                    result = f(map(value, inputs)...)
                    send_value!(output, result, timestep)
                    prev_timestep = timestep
                end
            end
        end
    end
end
```
`LATEST` is analogous to `ANY`. If this is no better than `eval`, then I can use `eval`.

**Comment by vtjnash** [comment]
Yes, I think longer-term we should do something like that. My latest attempt at defining something like this is https://discourse.julialang.org/t/proposal-for-a-first-class-dispatch-wrapper/1127/2

> Let's say that Reactive didn't launch its own event queue, and left that up to user code

Yeah, that wouldn't be good either. I think it would instead do something like have `add_action!` restart its event loop task. I don't know exactly what that would look like, but maybe something like:

```julia
__init__() = @async backend_eventqueue()
function backend_eventqueue()
    restart[] = false
    while true
        dispatch_events()
        restart[] && (@async backend_eventqueue(); return)
    end
end

**Pull request merged**